apps:healthz implement healthz URL

Madhu-1 opened the pull request:
<!--
Thanks for sending a pull request! Your contribution is appreciated.

Here are some tips for you:

1. Read the contributing guide at [https://github.com/heketi/heketi/blob/master/docs/contributing.md]
2. Split the changes up into minimal and atomic commits.  [https://github.com/heketi/heketi/blob/master/docs/contributing.md#splitting-your-change-into-commits]
3. Write meaningful commit messages [https://github.com/heketi/heketi/blob/master/docs/contributing.md#good-commit-messages]
4. Test your changes: run `make test`.  [https://github.com/heketi/heketi/blob/master/docs/contributing.md#testing-the-change]
5. Write a meaningful PR text. Remember: one PR per feature or bugfix. If in doubt, split your patchset into multiple PRs.
-->

### What does this PR achieve? Why do we need it?

`healthz` checks the following status in heketi.

* check DB is readable.
* check DB is writable.
* check the gluster nodes are reachable.

Signed-off-by: Madhu Rajanna <mrajanna@redhat.com>

### Does this PR fix issues?
Fixes #1412
<!-- This is optional, 'fixes #<issue-number>' lines will close the issue if the PR is merged.  -->

### Notes for the reviewer
TODO:
* split the commit in to multiple commits
* add test case

phlogistonjohn commented on apps/glusterfs/health.go:
```diff
@@ -0,0 +1,123 @@
+package glusterfs
```
This file looks to be missing the normal preamble we have in the source files in heketi.

Madhu-1 replied:
add license  path

phlogistonjohn commented on apps/glusterfs/health.go:
```diff
@@ -0,0 +1,123 @@
+	CurrentTime time.Time
+}
+
+func NewHealthz() *api.Healthz {
```
This function seems to be in an odd location between the type definition and that type's functions. Let's move it so it doesn't disrupt the flow when reading the code as much.

Madhu-1 replied:
removed

phlogistonjohn commented on apps/glusterfs/health.go:
```diff
@@ -0,0 +1,123 @@
+}
+
+func (*dbchecker) BucketName() string {
+	return BOLTDB_BUCKET_DBATTRIBUTE
```
If this is using the BOLTDB_BUCKET_DBATTRIBUTE bucket why isn't it using the DbAttributeEntry type? I really don't like the idea of multiple types sharing the same bucket. It will just lead to confusion IMO.

Madhu-1 replied:
create new db attribute for healthz

phlogistonjohn commented on apps/glusterfs/health.go:
```diff
@@ -0,0 +1,123 @@
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	//db Read check
+	err := a.db.View(func(tx *bolt.Tx) error {
```
All of the db checking logic should be encapsulated in it's own function call so that the actual rest endpoint function call is simply formatting and reporting on the status of the various checks but not doing all of them directly.

Madhu-1 replied:
done

phlogistonjohn commented on apps/glusterfs/health.go:
```diff
@@ -0,0 +1,123 @@
+	}
+	//cluster status check
+	err = a.db.View(func(tx *bolt.Tx) error {
+		health := NewNodeHealthCache(0, 0, a.db, a.executor)
```
Don't create a new health status cache. We should re-use the one previously set up when the app is set up. If the cache is disabled, simply report an empty but non-failure result.

Madhu-1 replied:
sorry I missed that we  are storing  the heathcache in app, updated code

phlogistonjohn commented on apps/glusterfs/health_cache.go:
```diff
@@ -171,6 +172,10 @@ func (hc *NodeHealthCache) toProbe() ([]*NodeHealthStatus, error) {
+
+	if !s.Up {
+		return
+	}
```
This doesn't make any sense to me. If the node is up and then goes down later how will ever detect that?

Madhu-1 replied:
updated code,  only if the node is marked as up, then only checking the glusterd status or else just  updating the state

phlogistonjohn replied:
I'm still confused. In addition this appears to be a conditional that does nothing. Is the current version complete?

phlogistonjohn commented on apps/glusterfs/health_cache.go:
```diff
@@ -154,14 +154,15 @@ func (hc *NodeHealthCache) toProbe() ([]*NodeHealthStatus, error) {
+			if !node.isOnline() {
+				continue
+			}
+			nhs.Up = true
```
I don't understand this change. What was wrong with the previous behavior? How does this change improve it? Because this is a subtle thing this is the kind of change I'd prefer to see in a separate commit with it's own commit message explaining the change (assuming it is still needed).

Madhu-1 replied:
i have added  it to todo list to make seperate commit,

currently, we  are saving only online node details  in the slice, for the health
implementation, we need  all nodes  status  irrespective of its state to decide whether
the cluster state is healthy or not.

phlogistonjohn replied:
Ok, makes sense. I do wonder if it would be beneficial to keeping two slices, one for up/healthy nodes and one for down/unhealthy nodes. That way when one needs only the list of healthy nodes code only needs to consult one slice w/o needing any additional conditionals.

phlogistonjohn commented on pkg/glusterfs/api/types.go:
```diff
@@ -693,3 +694,53 @@ func ValidateIds(v interface{}) error {
+	LastUpdate time.Time `json:"updated"`
+}
+
+type DbState uint8
```
I don't think this kind of detail is desirable in the api. Just have an enumerated string type with the ok/faulty info needed. That seems where you were going with the Healthz type, so I suggest simply removing this type from the api, and if you still think you want these types, locate them inside the app/glusterfs (for example).
The important part is not to put anything extra in the API.

Madhu-1 replied:
removed

phlogistonjohn commented on pkg/glusterfs/api/types.go:
```diff
@@ -693,3 +694,53 @@ func ValidateIds(v interface{}) error {
+	return state
+}
+
+type ClusterState uint8
```
See above.

phlogistonjohn commented on apps/glusterfs/healthz.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+func (*dbchecker) BucketName() string {
+	return BOLTDB_BUCKET_DBHEALTHZ
```
To be honest I'm really not sure we want a special bucket just for tihs, especially since we're saving a single value. Wouldn't it be simpler and less code to simply set/get an new key under dbattributes?

@raghavendra-talur : I'm curious if you have an opinion about this. The current approach is functional but feels a little like overkill to me. Could I get a 2nd opinion from you?

Madhu-1 replied:
@raghavendra-talur let me  know your thoughts on this

phlogistonjohn commented on apps/glusterfs/healthz.go:
```diff
@@ -0,0 +1,130 @@
+		if !n.Up {
+			totalUnhealthyNodes++
+		}
+		info.NodesStatus = append(info.NodesStatus, (api.NodeHealthStatus)(*n))
```
Are the parens around api.NodeHeathStatus needed? It makes it look like a C type cast but it doesn't look like idomatic go to me.

Madhu-1 replied:
yes this is needed for type casting

phlogistonjohn replied:
It's not. I just checked.

phlogistonjohn commented on apps/glusterfs/healthz.go:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	// cluster status check
+	checkClusterState(a.nhealth, &info)
```
I think this would read more logically to do the serialization once (at the bottom of the function) and move this line up into the if-staement (with an inverted condition) like so:
```golang
if MonitorGlusterNodes {
    checkClusterState(a.nhealth, &info)
}
// ... serialize response ...
```

phlogistonjohn commented:
@Madhu-1 do you plan to continue working on this PR?

Madhu-1 commented:
@phlogistonjohn I was waiting for @raghavendra-talur comments on nhttps://github.com/heketi/heketi/pull/1417#discussion_r258905434

raghavendra-talur commented:
@Madhu-1 I am not 100% clear about "check the gluster nodes are reachable" check.

- What is the need for the check
- Why are we looking at nodes without giving importance to cluster they belong?

raghavendra-talur commented:
> @Madhu-1 I am not 100% clear about "check the gluster nodes are reachable" check.
> 
> * What is the need for the check

I got the answer to this in the issue, the requirement is "Check that executor has been able to communicate with nodes in some timeframe"
> * Why are we looking at nodes without giving importance to cluster they belong?


I am not sure of the approach where:
1. none of the nodes are reachable == faulty
2. at least one of the nodes is reachable == healthy

Probably a timestamp of last successful reply(not at /healthz endpoint) would be a better candidate.

Madhu-1 commented:
@raghavendra-talur addressed review comments
 below is the response from curl request

```
curl http://127.0.0.01:8080/healthz |python -m json.tool
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   574  100   574    0     0  21259      0 --:--:-- --:--:-- --:--:-- 21259
{
    "cluster-state": "healthy",
    "db-state": "write",
    "nodes": [
        {
            "host": "192.168.10.101",
            "node": "252ced46d6cd6effb07a4688e72c829a",
            "online": true,
            "updated": "2019-03-27T15:59:15.063340783+05:30"
        },
        {
            "host": "192.168.10.102",
            "node": "ad4b1df157ac35018c6e9c21cb77ed5b",
            "online": true,
            "updated": "2019-03-27T15:59:15.4671574+05:30"
        },
        {
            "host": "192.168.10.100",
            "node": "b744aca95e4cd3b21ab6a92d497378f9",
            "online": false,
            "updated": "2019-03-27T15:59:18.559924198+05:30"
        },
        {
            "host": "192.168.10.103",
            "node": "d885c1bce818b58af247a7476b952209",
            "online": true,
            "updated": "2019-03-27T15:59:19.0096784+05:30"
        }
    ]
}
```

Madhu-1 commented:
> > @Madhu-1 I am not 100% clear about "check the gluster nodes are reachable" check.
> > 
> > * What is the need for the check
> 
> I got the answer to this in the issue, the requirement is "Check that executor has been able to communicate with nodes in some timeframe"
> 
> > * Why are we looking at nodes without giving importance to cluster they belong?
> 
> I am not sure of the approach where:
> 
> 1. none of the nodes are reachable == faulty

earlier approach was if a number of not reachable nodes are greater than the total nodes, I was marking cluster as faulty, now changed to code to make cluster state as faulty if all nodes are not reachable
> 2. at least one of the nodes is reachable == healthy
> 
> Probably a timestamp of last successful reply(not at /healthz endpoint) would be a better candidate.

raghavendra-talur commented:
@Madhu-1 @phlogistonjohn I am happy with the PR as it stands right now. I am however, not clear on what I want from /healthz in particular.

Should healthz be local to heketi's status(that means only the db read/write status)?
Or should healthz of heketi represent the status of the storage in whole?

phlogistonjohn commented:
I've looked over much of the code again and my comments and I stand by pretty much all of what I requested previously from my "requested changes". This is also going to need tests.

Madhu-1 commented:
@phlogistonjohn @raghavendra-talur currently I don't have enough bandwidth to work on this one, please feel free to take my patches if they are useful. Thank you :)

phlogistonjohn commented:
Effort is stale/abandoned. Closing.
